### PR TITLE
Tensorflow 1.x backend: add dropout to DeepONet

### PR DESCRIPTION
**Pull request by vl-dud**
Now DeepONet supports dropout technique.

### CODE REVIEW

**Comment by vl-dud** [comment]
> I am curious how useful it is. In general, I found dropout is not that useful, and L1/L2 regularization seems good enough.

Hyperparameter tuning showed that in my case neither dropout nor regularization is required. However I will use dropout anyway for UQ.

**Comment by lululxvi** [comment]
Yes, dropout is useful for UQ. How do you implement DeepONet UQ? Using the callback?

**Comment by vl-dud** [comment]
I just run `model.predict` many times to get the final prediction with CI:

```
def predict_with_uncertainty(model, x, trial_num=100):
    predictions = []
    for _ in range(trial_num):
        predictions.append(model.predict(x))
    return np.mean(predictions, axis=0), np.std(predictions, axis=0)
```

**Comment by lululxvi** [comment]
In fact, we have this callback https://deepxde.readthedocs.io/en/latest/modules/deepxde.html#deepxde.callbacks.DropoutUncertainty . Does this work for your case?

**Comment by vl-dud** [comment]
Yes, I used `DropoutUncertainty`. The above code snippet is more convenient in my case, since I use it with already trained models.
In addition, it allows you to set `trial_num`, which is a constant in `DropoutUncertainty`.

File: deepxde/nn/tensorflow_compat_v1/deeponet.py [hunk 1]
```diff
@@ -300,6 +303,10 @@ def build_branch_net(self):
                     activation=self.activation_branch,
                     trainable=self.trainable_branch,
                 )
+                if self.dropout_rate > 0:
+                    y_func = tf.layers.dropout(
+                        y_func, rate=self.dropout_rate, training=self.trainable_branch
```
Comment (lululxvi) [inline deepxde/nn/tensorflow_compat_v1/deeponet.py:308]: It should not be `self.trainable_branch`, which is used to indicate whether to train the branch during the training. It should be `self.training`, which indicates whether it is training or testing. See https://github.com/lululxvi/deepxde/blob/master/deepxde/nn/tensorflow_compat_v1/nn.py 
and
https://github.com/lululxvi/deepxde/blob/ad6399b27c3a0702abd24d45f889ca55497476c2/deepxde/callbacks.py#L307

Comment (vl-dud) [reply deepxde/nn/tensorflow_compat_v1/deeponet.py]: I fixed it. I suppose this applies to `self.trainable_trunk` as well.

File: deepxde/nn/tensorflow_compat_v1/deeponet.py [hunk 2]
```diff
@@ -553,6 +574,8 @@ def build_branch_net(self):
                     kernel_initializer=self.kernel_initializer,
                     kernel_regularizer=self.regularizer,
                 )
+                if self.dropout_rate > 0:
+                    y_func = tf.layers.dropout(y_func, rate=self.dropout_rate)
```
Comment (lululxvi) [inline deepxde/nn/tensorflow_compat_v1/deeponet.py:578]: Miss `training=self.training`

File: deepxde/nn/tensorflow_compat_v1/deeponet.py [hunk 3]
```diff
@@ -331,15 +342,22 @@ def build_trunk_net(self):
...
+            )
             y_loc = self._dense(
                 y_loc,
                 self.layer_size_loc[i],
                 activation=self.activation_trunk,
                 regularizer=self.regularizer,
-                trainable=self.trainable_trunk[i - 1]
-                if isinstance(self.trainable_trunk, (list, tuple))
-                else self.trainable_trunk,
+                trainable=trainable,
```
[the remark this answers is not in the document]
Comment (vl-dud) [reply deepxde/nn/tensorflow_compat_v1/deeponet.py:355]: I assumed that the same logic applies here as in the branch network. I am wrong?

File: deepxde/nn/tensorflow_compat_v1/deeponet.py (hunk 2 above)
Comment (vl-dud) [reply deepxde/nn/tensorflow_compat_v1/deeponet.py:578]: For class `DeepONetCartesianProd`, `self.training` is not used anywhere. Should I add it only on line 578?

File: deepxde/nn/tensorflow_compat_v1/deeponet.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lululxvi) [reply deepxde/nn/tensorflow_compat_v1/deeponet.py:355]: I mean the new code (line 355, line 345--349) works identically as the original code.

File: deepxde/nn/tensorflow_compat_v1/deeponet.py (hunk 2 above)
Comment (lululxvi) [reply deepxde/nn/tensorflow_compat_v1/deeponet.py:578]: I guess so. In `DeepONet`, `self.training` is also only used a few times, right? But it is better to test the code.

Comment (vl-dud) [reply deepxde/nn/tensorflow_compat_v1/deeponet.py]: Fixed

File: deepxde/nn/tensorflow_compat_v1/deeponet.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vl-dud) [reply deepxde/nn/tensorflow_compat_v1/deeponet.py]: In the previous [commit](https://github.com/lululxvi/deepxde/pull/1579/commits/71f1a460662b904eedf116d4c2d5db3cea57e8fb), the `trainable` variable is used twice. Now this is unnecessary. I fixed it.
